python/cudf/cudf/utils/performance_tracking.py

```diff
@@ -0,0 +1,80 @@
+# Copyright (c) 2023-2024, NVIDIA CORPORATION.
+
+import contextlib
+import functools
+import hashlib
+import sys
+from typing import Dict
+
+import nvtx
+
+import rmm.statistics
+
+from cudf.options import get_option
+
+_NVTX_COLORS = ["green", "blue", "purple", "rapids"]
+
+
+def _get_color_for_nvtx(name):
+    m = hashlib.sha256()
+    m.update(name.encode())
+    hash_value = int(m.hexdigest(), 16)
+    idx = hash_value % len(_NVTX_COLORS)
+    return _NVTX_COLORS[idx]
+
+
+def _performance_tracking(func, domain="cudf_python"):
+    """Decorator for applying performance tracking (if enabled)."""
+
+    @functools.wraps(func)
+    def wrapper(*args, **kwargs):
+        with contextlib.ExitStack() as stack:
+            if get_option("memory_profiling"):
+                rmm.statistics.enable_statistics()
```

Comment:
*issue*: This doesn't behave like a normal context manager (leave the state unchanged after exit) because it doesn't unwind the change to the current memory resource that `rmm.statistics.enable_statistics` enacts. This means that the following code does an "unexpected" thing: side-effectfully changes the MR:

```python
# statistics not enabled...
with cudf.options_context(("memory_profiling", True)):
    # code that profiles statistics

# expect stats memory resource to be popped, but it is not
```

It seems like the call to enable statistics should be paired with a matching disable statistics call. But this seems bad if it were done at every level of the call stack.

It _feels_ like the right place for modification of the MR is in the `set_option` call. `set_option("memory_profiling", True)` would store the old MR, push the statistics MR, and then a later `set_option("memory_profiling", False")` would pop the statistics MR if it still matches.

This, however, would preclude use of the environment variable default (because that is set on import, and we don't want to have to call RMM on import).

Does this suggest that we need a specific `enable_memory_profiling` that can be used as a context manager that sets the option and sets up the MR?

Reply:
If this is an issue, RMM already have `rmm.statistics.statistics()` and `rmm.statistics.enable_statistics()`.
Thus the user code would look something like:
```python
import cudf
import rmm.statistics

# The must first enable the cudf option
cudf.set_option("memory_profiling", True)

# Then enable statistics within a context
with rmm.statistics.statistics():
    # by code

# Or globally
rmm.statistics.enable_statistics()
```
Of cause, we could wrap/alias the rmm functions in cudf native functions. However, I am not sure if the disadvantage of the "unexpected" side-effect outweigh the user-friendliness of just setting `CUDF_MEMORY_PROFILING="1"`?

Reply:
Alternatively, we could implement lazy environment variable in `options.py`. It might be worth streamline the option module in any case.

Reply:
In terms of solutions, it depends a little bit on how end-user-friendly we want this to be. If this is mostly for developer use, I am happy with the programmatic set_option + use RMM calls. If we want this as a debugging tool for end users "where is my code allocating lots of vram", then we likely need something a little more ergonomic.

As much as I am allergic to environment variables from a reproducibility point of view, I can see the attraction.

Reply:
I am not sure but I guess we could start with the `set_option + use RMM calls` and then later consider a more end-user-friendly solution?

Reply:
I think that's a reasonable compromise.

Reply:
done